ntoskrnl/lpc/reply.c

```diff
@@ -754,15 +847,278 @@ NtReplyWaitReceivePort(IN HANDLE PortHandle,
+    if (!(WakeupThread->LpcExitThreadCalled) &&
+        !(IsListEmpty(&WakeupThread->LpcReplyChain)))
+    {
+        /* Remove us from it and re-initialize it */
+        RemoveEntryList(&WakeupThread->LpcReplyChain);
+        InitializeListHead(&WakeupThread->LpcReplyChain);
+    }
+
+    /* Set us up to get the following reply */
+    Thread->LpcReplyMessageId = CapturedReplyMessage.MessageId;
+    Thread->LpcReplyMessage = NULL;
+
+    /* Check if this is the message the thread has received */
+    if ((Thread->LpcReceivedMsgIdValid) &&
+        (Thread->LpcReceivedMessageId == CapturedReplyMessage.MessageId))
+    {
+        /* Clear this data */
+        Thread->LpcReceivedMessageId = 0;
+        Thread->LpcReceivedMsgIdValid = FALSE;
+    }
+
+    /* Insert the thread into the rundown queue */
+    if ((Port->Flags & LPCP_PORT_TYPE_MASK) != LPCP_CONNECTION_PORT)
+    {
+        RundownPort = Port->ConnectedPort;
+    }
+    else
+    {
+        RundownPort = Port;
+    }
+
+    /* Insert the message in our chain */
+    InsertTailList(&RundownPort->LpcReplyChainHead, &Thread->LpcReplyChain);
+    LpcpSetPortToThread(Thread, Port);
+
+    /* Free any data information */
+    LpcpFreeDataInfoMessage(Port,
+                            CapturedReplyMessage.MessageId,
+                            CapturedReplyMessage.CallbackId,
+                            CapturedReplyMessage.ClientId);
+
+    /* Release the lock and release the LPC semaphore to wake up waiters */
+    KeReleaseGuardedMutex(&LpcpLock);
+    LpcpCompleteWait(&WakeupThread->LpcReplySemaphore);
+
+    /* Now we can let go of the thread */
+    ObDereferenceObject(WakeupThread);
+
+    /* Now wait for a reply and set 'Status' */
+    LpcpConnectWait(&Thread->LpcReplySemaphore, PreviousMode);
+
+    /* Acquire the lock */
+    KeAcquireGuardedMutex(&LpcpLock);
+
+    /* Check if the reply chain is not empty */
+    if (!IsListEmpty(&Thread->LpcReplyChain))
+    {
+        /* Remove this entry and re-initialize it */
+        RemoveEntryList(&Thread->LpcReplyChain);
+        InitializeListHead(&Thread->LpcReplyChain);
+    }
+
+    /* Check if wait succeeded */
+    if (Status == STATUS_SUCCESS)
+    {
+		/* Get the LPC message and clear our thread's reply data */
+        Message = LpcpGetMessageFromThread(Thread);
+        Thread->LpcReplyMessage = NULL;
+
+        /* Release the lock */
+        KeReleaseGuardedMutex(&LpcpLock);
+
+        /* Check if we have a valid message */
+        if (Message != NULL)
+        {
+            /* Copy the message */
+            _SEH2_TRY
+            {
+                LpcpMoveMessage(ReplyMessage,
+                                &Message->Request,
+                                (&Message->Request) + 1,
+                                0,
+                                NULL);
+            }
+            _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+            {
+                DPRINT1("Got exception!\n");
```

Comment:
Useless unless you also print the exception code.

Reply:
Right.